File: Dec2020/3Dec/KthMissingPositiveNumber.java
/*
Problem Name: 1539. Kth Missing Positive Number
Problem Link: https://leetcode.com/problems/kth-missing-positive-number/
*/
class Solution {
    public int findKthPositive(int[] arr, int k) {
        int cnt[] = new int[2500];
        for(int i=0;i<arr.length;i++)
            cnt[arr[i]] = 1;
        for(int i=1;k>0;i++){
            if(cnt[i] == 0){
                k--;
            }
            if(k == 0)
                return i;
        }
        return 0;
    }
}

File: Dec2020/3Dec/KthMissingPositiveNumber.cpp
/*
Problem name: 1539. Kth Missing Positive Number
Problem link: https://leetcode.com/problems/kth-missing-positive-number/
*/

class Solution {
public:
    int findKthPositive(vector<int>& arr, int k) {
        if(arr[0] - 1 >= k)
            return k;
        
        k -= arr[0] - 1;
        int d, i;
        
        for(i = 1; i < arr.size(); i++){
            d = arr[i] - arr[i-1] - 1;
            
            if(d == 0)
                continue;
            else if(d >= k)
                return arr[i-1] + k;
            else
             k -= d;                
        }
        return arr[i - 1] + k;
    }
};
